RecordInvalid 時の例外処理の実装

chionyan opened the pull request:
このPRにおける目的
----
- [RecordInvalid 例外処理の入力値に対する返却値 · Issue \#18 · chionyan/simple\-todo\-api](https://github.com/chionyan/simple-todo-api/issues/18#issuecomment-416460609)に従い`POST /todos` `PATCH /todos/:id` における Validation error 時の例外処理を実装する

やったこと
----
- `ActionController:: RecordInvalid ` の例外処理実装

特記事項
----
- リクエストボディは `{ "title": "...", "text": "..." }` の JSON 形式に限定する。

chionyan committed:
Add RecordInvalid Feature & Test

chionyan commented:
(口頭説明)
[MakeShop 初回連携 API](https://github.com/feedforce/dfplusio-api/wiki/%E4%BB%95%E6%A7%98%EF%BC%9AMakeShop-%E9%80%A3%E6%90%BA%E6%A9%9F%E8%83%BD#user-content-makeshop-%E5%88%9D%E5%9B%9E%E9%80%A3%E6%90%BA-api)のように
```
POST "http://localhost:3000/todos"
{
  "data": {
    "attributes": {
      "title": "タイトル",
      "text": "テキスト"
    }
  }
}
```
をリクエストパラメータとして送った結果、 `"pointer": "/data/attributes/{attribute name}"` が返ってくるものだと考えられる。
したがって、リクエストパラメータとして送るデータの階層が異なると、 `pointer` として返ってくる階層も違ってくる。

chionyan committed:
Merge branch 'master' into error-invalid

chionyan committed:
Merge & Conflict resolution

chionyan committed:
Fix RecordInvalid Error Output

chionyan commented:
## 動作確認
~~4474977~~ 10f2f6d での修正に当たって、出力結果を修正しています。

- [ ] Rails
  - `$ bin/setup` を実行
  - `$ bundle exec rails s` で Rails サーバーを起動

  ### POST /todos

  - [ ] `title` が空文字の時
  ```sh
  $ curl -X POST "http://localhost:3000/todos" -H "accept: application/json" -H "Content-Type: application/json" -d '{"title":"", "text": "Sample text 1"}' | jq
  {
    "errors": [
      {
        "title": "タイトルを入力してください'",
        "status": 422,
        "source": {
          "pointer": "/title"
        }
      }
    ]
  }
  ```

  - [ ] `text` が空文字の時
  ```sh
  $ curl -X POST "http://localhost:3000/todos" -H "accept: application/json" -H "Content-Type: application/json" -d '{"title":"Sample title 1", "text": ""}' | jq
  {
    "errors": [
      {
        "title": "テキストを入力してください'",
        "status": 422,
        "source": {
          "pointer": "/text"
        }
      }
    ]
  }
  ```

  - [ ] `title` `text` が空文字の時
  ```sh
  $ curl -X POST "http://localhost:3000/todos" -H "accept: application/json" -H "Content-Type: application/json" -d '{"title":"", "text": ""}' | jq
  {
    "errors": [
      {
        "title": "タイトルを入力してください'",
        "status": 422,
        "source": {
          "pointer": "/title"
        }
      },
      {
        "title": "テキストを入力してください'",
        "status": 422,
        "source": {
          "pointer": "/text"
        }
      }
    ]
  }
  ```

  ### PATCH /todos/:id

  - 一件レコード作成
  ```sh
  $ curl -X POST "http://localhost:3000/todos" -H "accept: application/json" -H "Content-Type: application/json" -d '{"title":"Sample title 1", "text": "Sample text 1"}' | jq
  ```

  - [ ] `title` が空文字の時
  ```sh
  $ curl -X PATCH "http://localhost:3000/todos/:id" -H "accept: application/json" -H "Content-Type: application/json" -d '{"title":"", "text": "Change text 1"}' | jq
  {
    "errors": [
      {
        "title": "タイトルを入力してください'",
        "status": 422,
        "source": {
          "pointer": "/title"
        }
      }
    ]
  }
  ```

  - [ ] `text` が空文字の時
  ```sh
  $ curl -X PATCH "http://localhost:3000/todos/:id" -H "accept: application/json" -H "Content-Type: application/json" -d '{"title":"Change title 1", "text": ""}' | jq
  {
    "errors": [
      {
        "title": "テキストを入力してください'",
        "status": 422,
        "source": {
          "pointer": "/text"
        }
      }
    ]
  }
  ```

  - [ ] `title` `text` が空文字の時
  ```sh
  $ curl -X PATCH "http://localhost:3000/todos/:id" -H "accept: application/json" -H "Content-Type: application/json" -d '{"title":"", "text": ""}' | jq
  {
    "errors": [
      {
        "title": "タイトルを入力してください",
        "status": 422,
        "source": {
          "pointer": "/title"
        }
      },
      {
        "title": "テキストを入力してください'",
        "status": 422,
        "source": {
          "pointer": "/text"
        }
      }
    ]
  }
  ```

- [ ] RSpec
  - `$ bundle exec rspec` を実行
  - [ ] `0 failures` が出力されることを確認

- [ ] Rubocop
  - `$ bundle exec rubocop` を実行
  - [ ] `no offenses detected` が出力されることを確認

chionyan commented:
@koshigoe @pokotyamu 
`todos#create` `todos#update` におけるバリデーションエラーの例外処理のテストのPRを作成しました！レビューお願いします🙏
実装における決定事項などは、[Issue \#18](https://github.com/chionyan/simple-todo-api/issues/18#issuecomment-416460609)に記述していますので、一度目を通していただけるとと思います🙇‍♀️

koshigoe commented:
👁

koshigoe commented on app/controllers/todos_controller.rb:
```diff
@@ -6,6 +6,15 @@ class TodosController < ApplicationController
+    end
+    render json: { errors: errors }, status: 422
+  end
+
```
❓バリデーション違反の原因に関わらず、 `/errors/i/title` の値が `"バリデーションに失敗しました。"` 固定である理由を知っておきたいです。利用者としては、具体的にどういう理由で受け入れられなかったのかを知りたいかと思うのですが、敢えて詳細な情報は伝えないポリシーという事でしょうか？

koshigoe replied:
❓バリデーション違反のメッセージを使わないのであれば、[`ActiveModel::Errors#keys`](https://api.rubyonrails.org/classes/ActiveModel/Errors.html#method-i-keys)を使えば良さそうに見えますが、そんなことないですかね？

koshigoe replied:
❓余談(感想)ですが `map` を使えばブロックの中で `<<` する必要がなくなるかなと思いました。

koshigoe replied:
❓これも余談ですが、 キー `'pointer'` はシンボルでなく文字列なんですね。

koshigoe replied:
❓さらに感想ですが、「バリデーションに失敗しました」という日本語を読んで、バリデーション処理が正常に行えなかったのか、入力値がバリデーションルールに違反していたのか、よく分からないなと思いました。が、これは、自分の日本語能力の問題な気もします。

chionyan replied:
> 余談(感想)ですが map を使えばブロックの中で << する必要がなくなるかなと思いました。

🅰️ こちらはうまいこと書き換えれそうな気がしますので、修正してみます！

> これも余談ですが、 キー 'pointer' はシンボルでなく文字列なんですね。

🅰️ こちらはうっかりミスだったので修正します😂

> バリデーション違反のメッセージが固定である件について

🅰️ Swagger の仕様で `"title": "..."` となっていて何を入れたらいいかわからなかったので、とりあえず固定のメッセージを入れていました。ただ実装しながら、コメントいただいた通り、自分がユーザ目線だった場合に何がエラーなのかわからないので、このPR内でエラーメッセージを具体的にする修正を行います❗️

chionyan replied:
バリデーション違反のメッセージについては、 eef84da で ActiveRecord の日本語辞書を追加し、 
4474977 で詳細なメッセージを表示しています！
`pointer` のシンボルの件や、 `#map` 使って配列を書き換える修正も 4474977 で行なっています！
ついでに `i18n` の翻訳で行が長くなることなどから、メソッドを切り出して、 NotFound の方も 
91c22b6 で修正しています✨

koshigoe reviewed (commented):
取り急ぎ、ざっくりとした感想まで。

chionyan committed:
Add Activerecord Dictionary

chionyan committed:
Fix ActiveRecord::RecordInvalid Error Process

- ActiveRecord の辞書を追加（エラーメッセージを詳細にするため）
- 翻訳メソッドを切り出し

chionyan committed:
Refactor ActiveRecord::RecordNotFound Error Process

koshigoe commented on app/controllers/todos_controller.rb:
```diff
@@ -2,10 +2,27 @@ class TodosController < ApplicationController
+    end
+    render json: { errors: errors }, status: 422
+  end
+
```
❓ ActiveModel のバリデーション機能で得られるエラーメッセージを使っていないのは意図しての事ですか？

> i18n の翻訳で行が長くなることなどから、メソッドを切り出し

`ActiveModel::Errors` によるメッセージの組み立てで内部的に `I18n` が使われている事はご存じです？

- :bookmark: https://guides.rubyonrails.org/i18n.html#translations-for-active-record-models
- :bookmark: https://github.com/rails/rails/blob/184c1a405f58b069e0b9ef79a2ad67f7d7d9e69b/activemodel/lib/active_model/errors.rb
- :bookmark: https://github.com/rails/rails/blob/184c1a405f58b069e0b9ef79a2ad67f7d7d9e69b/activemodel/lib/active_model/locale/en.yml
- :bookmark: https://github.com/svenfuchs/rails-i18n/blob/2fbcb49765d09b17ccee204be96b17ff888beeec/rails/locale/ja.yml

```
5.2.1@2.5.0 (main)> todo = Todo.new
=> #<Todo:0x00007fd5ca9b4d38 id: nil, title: nil, text: nil, created_at: nil, updated_at: nil>
5.2.1@2.5.0 (main)> todo.validate
=> false
5.2.1@2.5.0 (main)> todo.errors.full_messages
=> ["Title can't be blank", "Text can't be blank"]
5.2.1@2.5.0 (main)> todo.errors.messages
=> {:title=>["can't be blank"], :text=>["can't be blank"]}
5.2.1@2.5.0 (main)> I18n.with_locale(:ja) { todo.errors.full_messages }
=> ["タイトル can't be blank", "テキスト can't be blank"]
5.2.1@2.5.0 (main)> I18n.with_locale(:ja) { todo.errors.messages }
=> {:title=>["can't be blank"], :text=>["can't be blank"]}
```

koshigoe replied:
> e.record.errors.messages.keys

🤔 バリデーション違反の属性名が必要なだけであれば `ActiveModel::Errors#message` を経由せず `ActiveModel::Errors#keys` で良かったりしませんか？(https://github.com/chionyan/simple-todo-api/pull/16#discussion_r213204449)

> errors.map!

🤔 変数 `errors` に代入してから `map!` するくらいなら、 `errors = e.record.errors.keys.map do |attribute|` で良かったりしませんか？

chionyan replied:
> ActiveModel のバリデーション機能で得られるエラーメッセージを使っていないのは意図しての事ですか？
> ActiveModel::Errors によるメッセージの組み立てで内部的に I18n が使われている事はご存じです？

🅰️ こちらについては、 gem `rails-i18n` をインストールし `default_locale = :ja` とした上で、`ActiveModel::Errors` からメッセージを取得します！

chionyan replied:
- `rails-i18n` のインストールは f721e5a 
- `default_locale = :ja` の設定は 548a666 
- `errors` の組み立てについては 10f2f6d
で修正しております✨

koshigoe reviewed (commented):
改めてコメントしました。

質問について、意図的な仕様だったり、意識的なスタイルだ、という事であれば特に修正せずそのままで良いかと思いました。

chionyan committed:
Install rails-i18n

chionyan committed:
Setting rails-i18n default_locale = :ja

chionyan committed:
Fix ActiveRecord::RecordInvalid Error Process2

- ActiveModel::Errors#full_messages を使用
- locale の指定を削除
- 切り分けたメソッドを削除
- テストの修正

chionyan commented:
@koshigoe 
指摘事項の修正が完了しましたので、再度レビューお願いします🙏
@pokotyamu 
昨日少し見ていただいた部分からちょっと変更点などもあるので（切り出したメソッドの削除など）、もう一度見ていただいて何かあればご指摘お願いします✨

chionyan committed:
Merge branch 'master' into error-invalid

koshigoe commented on app/controllers/todos_controller.rb:
```diff
@@ -2,10 +2,28 @@ class TodosController < ApplicationController
+
+    render json: { errors: errors }, status: 422
+  end
+
```
任意の属性に関するバリデーションメッセージを取得する場合、属性名を指定すべきだと思います。

- [`ActiveModel::Errors#full_messages_for`](https://api.rubyonrails.org/classes/ActiveModel/Errors.html#method-i-full_messages_for)

`ActiveModel::Errors#full_messages[i]` がどの属性のメッセージを意味するかは、 `ActiveModel::Errors#keys` のインデックスから安定的に特定可能ではないと思います。
ActiveModel の仕組み上、ある属性に対するバリデーションエラーメッセージはひとつだとは限らないため、その仕組みを前提としたロジックであるべきだというのが、私の考えです。

chionyan replied:
コメントありがとうございます✨
`ActiveModel::Errors#full_messages_for` の存在を知らなかったので、インデックスを使うより属性を使って取り出した方が確実なメッセージを取り出せるとわかりました！
ac25274 で `ActiveModel::Errors#full_messages_for(attribute)` を使用してメッセージを取り出す方法に修正しました！ `full_messages_for` の戻り値のクラスが `Array` なのでそちらも分解して取り出せるように `#map` で回してます！

koshigoe reviewed (changes requested):
バリデーションのエラーメッセージを取り出すロジックについて、コメントしました。

chionyan committed:
Fix Create Error JSON Process

koshigoe reviewed (approved):
ループのネストが若干気になるけど、代替案がぱっと思いつかない。

仕組みとしては問題ないかなと思いました。
